### PR TITLE
Implement new family of stack functions

### PR DESCRIPTION
**Pull request by lionel-**
Todo: a vignette on stacks and execution contexts.

I'm thinking maybe we should rename all `ctxt_` functions to the `frame_` prefix? Because "context" has the potential to be a bit confusing as the R-level contexts are a subset of C-level contexts. So we could talk about the frame stack instead, with frame standing for evaluation environment (which are the only contexts exposed to R).

### CODE REVIEW

**Comment by wch** [comment]
A couple of thoughts:

* In the first paragraph, it would be helpful if the last sentence (about the base functions that give contexts vs. call stack) were earlier. I think moving it to the second sentence is the right place.
* In the vignette, it would help to show some examples where the eval context differs from the call stack. An example that shows how things could break if you expect them to be the same.

Here's a possible example:

```R
wrapper <- function(x) {
  y <- 2
  x
}

get_var_call_stack <- function(name) {
  parent.frame()[[name]]
}

get_var_context_stack <- function(name) {
  sys.frame(-1)[[name]]
}


y <- 1

get_var_call_stack("y")
wrapper(get_var_call_stack("y"))

get_var_context_stack("y")
wrapper(get_var_context_stack("y"))
```

**Comment by hadley** [comment]
Is it necessary to have to so many functions? Could we reduce the number by having them return a more complex object?

**Comment by lionel-** [comment]
thanks @wch 

@hadley There is already `ctxt_stack()` and `call_stack()` which return a list of frame objects. I have now unexported the plural functions and gathered the singular functions to `ctxt_frame()` and `call_frame()`.

About the names, an alternative for `ctxt_` is `eval_`. So we'd have the C-level context stack, and the evaluation and call stacks.

File: R/stack.R [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+#' \code{which} argument are more complicated and changes meaning for
+#' values lower than 1. For the sake of consistency, the lazyeval
+#' functions all take the same kind of argument \code{n}. This
+#' argument has a single meaning (the number of frames to go back in
+#' the stack) and cannot be lower than 1.
+#'
+#' \itemize{
+#'   \item The \code{_expr()}, \code{_env()}, \code{_caller()} and
+#'         \code{_fun()} functions report about the call expression,
+#'         the calling environment, and the caller function
```
Comment (lionel-) [inline R/stack.R]: caller is about the position of the calling frame, while fun is the closure of the current frame. Should it be `caller_pos`?

File: R/stack.R [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+#' execution history but due to lazy evaluation it is potentially
+#' nonlinear in terms of call history. The call stack on the other
+#' hand is homogeneous in this regard. See the vignette on execution
+#' contexts for more information. Note that the base \code{sys.___()}
+#' functions are about the context stack and the only base R function
+#' for querying the call stack is \code{\link{parent.frame}()}.
+#'
+#' The base R functions take two sorts of arguments to indicate which
+#' frame to query: \code{which} and \code{n}. The \code{n} argument is
+#' straightforward: it's the number of frames to go back in the stack,
```
Comment (hadley) [inline R/stack.R]: Yes, I think so.

File: R/stack.R (hunk 1 above)
Comment (hadley) [inline R/stack.R]: I think it's ok as is, but you should mention it in the text-comma-list

Comment (lionel-) [inline R/stack.R]: I've renamed it to `caller_pos` so it's consistent with `pos`.

**Comment by lionel-** [comment]
Merged in rlang.

**Pull request closed without merging**